Add Retry Events

hinthornw opened the pull request:
![image](https://github.com/hwchase17/langchain/assets/13333726/59a5c3b4-4367-47e6-9f58-5b6557576a8a)

hinthornw committed:
tmp

hinthornw committed:
Dedup methods

hinthornw committed:
Add retry event

hinthornw committed:
str

nfcampos commented on langchain/callbacks/manager.py:
```diff
@@ -571,6 +574,22 @@ def on_llm_new_token(
 
+    def on_llm_retry(
+        self,
+        retry_state: RetryCallState,
```
Can we make this arg something that is json serialisable? Analysing that python repl string wont be fun

hinthornw replied:
Ya I can take out the useful bits

nfcampos commented on langchain/callbacks/manager.py:
```diff
@@ -571,6 +574,22 @@ def on_llm_new_token(
             **kwargs,
         )
 
+    def on_llm_retry(
```
why only llms? it could just be on_retry and be available in all run types?

nfcampos commented on langchain/callbacks/tracers/base.py:
```diff
@@ -7,9 +7,11 @@
 from langchain.callbacks.base import BaseCallbackHandler
 from langchain.callbacks.tracers.schemas import Run, RunTypeEnum
-from langchain.load.dump import dumpd
+from langchain.load.dump import dumpd, dumps
```
I dont see where this is being used?

hinthornw committed:
merge

hinthornw committed:
generic retry event

hinthornw committed:
move to base

hinthornw committed:
fix import

hinthornw committed:
change obj

hinthornw committed:
update

leo-gan commented:
Unit tests would be nice for this important code.

hinthornw committed:
format

hinthornw committed:
Add couple of tests

hinthornw committed:
Merge branch 'master' into wfh/retry_events

hinthornw committed:
Move retries to int test, update to smoke test

nfcampos commented on libs/langchain/langchain/callbacks/tracers/base.py:
```diff
@@ -138,6 +140,41 @@ def on_llm_new_token(
+            retry_d["outcome"] = "failed"
+            exception = retry_state.outcome.exception()
+            retry_d["exception"] = str(exception)
+            retry_d["exception_cls"] = exception.__class__.__name__
```
`cls` is a python thing, can we do something more cross-platform-y?

hinthornw replied:
`exception_type` perhaps?

hinthornw committed:
rename

ruoccofabrizio committed:
Azure Cognitive Search: Custom index and scoring profile support (#6843)

Description: Adding support for custom index and scoring profile support
in Azure Cognitive Search
@hwchase17

---------

Co-authored-by: Bagatur <baskaryan@gmail.com>

misrasaurabh1 committed:
Optimize the cosine_similarity_top_k function performance (#8151)

Optimizing important numerical code and making it run faster.

Performance went up by 1.48x (148%). Runtime went down from 138715us to
56020us

Optimization explanation:

The `cosine_similarity_top_k` function is where we made the most
significant optimizations.
Instead of sorting the entire score_array which needs considering all
elements, `np.argpartition` is utilized to find the top_k largest scores
indices, this operation has a time complexity of O(n), higher
performance than sorting. Remember, `np.argpartition` doesn't guarantee
the order of the values. So we need to use argsort() to get the indices
that would sort our top-k values after partitioning, which is much more
efficient because it only sorts the top-K elements, not the entire
array. Then to get the row and column indices of sorted top_k scores in
the original score array, we use `np.unravel_index`. This operation is
more efficient and cleaner than a list comprehension.

The code has been tested for correctness by running the following
snippet on both the original function and the optimized function and
averaged over 5 times.
```
def test_cosine_similarity_top_k_large_matrices():
    X = np.random.rand(1000, 1000)
    Y = np.random.rand(1000, 1000)
    top_k = 100
    score_threshold = 0.5
    gc.disable()
    counter = time.perf_counter_ns()
    return_value = cosine_similarity_top_k(X, Y, top_k, score_threshold)
    duration = time.perf_counter_ns() - counter
    gc.enable()
```

@hwaking @hwchase17 @jerwelborn 

Unit tests pass, I also generated more regression tests which all
passed.

baskaryan committed:
merge

baskaryan commented:
ik jinachat, cohere, and vertexai (and probs more) have v similar retry logic, should we add there, too? or hold off for sep pr?

hinthornw committed:
Delete Old Evals Examples (#8252)

Still retain:
- Comparison Examples
- Data + QA walkthrough
- QA (but really minimize it)

hinthornw commented:
I think probably separate PR. I'll fix the merge conflicts though

hinthornw committed:
emb

hinthornw committed:
Merge branch 'master' into wfh/retry_events
